shorebird_code_push/lib/src/shorebird_code_push.dart

```diff
@@ -2,48 +2,101 @@ import 'dart:isolate';
 
 import 'package:shorebird_code_push/src/updater.dart';
 
+/// A function that constructs an [Updater] instance. Used for testing.
+typedef UpdaterBuilder = Updater Function();
+
+/// A logging function for errors arising from interacting with the native code.
+///
+/// Used to override the default behavior of using [print].
+typedef ShorebirdLog = void Function(Object? object);
+
 /// {@template shorebird_code_push}
 /// Get info about your Shorebird code push app.
 /// {@endtemplate}
 class ShorebirdCodePush {
   /// {@macro shorebird_code_push}
   ShorebirdCodePush({
-    Updater Function()? createUpdater, // for testing
-  }) : _createUpdater = createUpdater ?? Updater.new;
+    this.logError = print,
+    UpdaterBuilder? buildUpdater, // for testing
+  }) : _buildUpdater = buildUpdater ?? Updater.new;
+
+  /// Logs error messages arising from interacting with the native code.
+  ///
+  /// Defaults to [print].
+  final ShorebirdLog logError;
 
-  final Updater Function() _createUpdater;
+  final UpdaterBuilder _buildUpdater;
+  static const _loggingPrefix = '[ShorebirdCodePush]';
 
   /// Checks whether a new patch is available for download.
   ///
   /// Runs in a separate isolate to avoid blocking the UI thread.
   Future<bool> checkForUpdate() {
-    return _runInIsolate((updater) => updater.checkForUpdate());
+    return _runInIsolate(
+      (updater) => updater.checkForUpdate(),
+      fallbackValue: false,
+    );
   }
 
   /// The version of the currently-installed patch. Null if no patch is
   /// installed (i.e., the app is running the release version).
   Future<int?> currentPatchNumber() {
-    return _runInIsolate((updater) {
-      final patchNumber = updater.currentPatchNumber();
-      return patchNumber == 0 ? null : patchNumber;
-    });
+    return _runInIsolate(
+      (updater) {
+        final patchNumber = updater.currentPatchNumber();
+        return patchNumber == 0 ? null : patchNumber;
+      },
+      fallbackValue: null,
+    );
   }
 
   /// The version of the patch that will be run on the next app launch. If no
   /// new patch has been downloaded, this will be the same as
   /// [currentPatchNumber].
   Future<int?> nextPatchNumber() {
-    return _runInIsolate((updater) {
-      final patchNumber = updater.nextPatchNumber();
-      return patchNumber == 0 ? null : patchNumber;
-    });
+    return _runInIsolate(
+      (updater) {
+        final patchNumber = updater.nextPatchNumber();
+        return patchNumber == 0 ? null : patchNumber;
+      },
+      fallbackValue: null,
+    );
+  }
+
+  /// Downloads the latest patch, if available.
+  Future<void> downloadUpdate() async {
```

Comment:
Would be nice to return a Stream of progress updates eventually.

Reply:
https://github.com/shorebirdtech/updater/issues/38 :)